File: CivitasRuby/JugadorEspeculador.rb
module Civitas
    class JugadorEspeculador < Jugador

        @@FactorEspeculador=2
        @@CasasMax=@@FactorEspeculador*@@CasasMax
        @@HotelesMax=@@HotelesMax*@@FactorEspeculador

        def initialize(fianza)
            
            @fianza=fianza
            
        end

        private_class_method :new

        def self.nuevoEspeculador(jugador,fianza)

            nuevo=new(fianza)
            nuevo.copia(jugador)

            for i in (0...nuevo.propiedades.size-1)
                nuevo.propiedades[i].actualizaPropietarioPorConversion(nuevo)
            end

            return nuevo
            
        end

        def pagaImpuesto(cantidad)
        
            super(cantidad/2)
        end

        def debeSerEncarcelado 
            @encarcelado ? ret=false : ret= true 

            if !@encarcelado && !tieneSalvoConducto && saldo<= @fianza
                ret=true
            else 
                if tieneSalvoConducto
                    perderSalvoConducto
                    Diario.instance.ocurre_evento('El jugador ' + @nombre + ' se libra de la carcel.')
                    ret=false
                else
                    paga(@fianza)
                end
            end

            return ret
        end

        def getCasasMax
            @@CasasMax
        end

        def getHotelesMax
            @@HotelesMax
        end

        def toString

            salida = super
            salida+= 'Es un jugador especulador' + "\n"

            return salida
        end

    end
end

File: CivitasJava/civitas/JugadorEspeculador.java
package civitas;

import java.util.ArrayList;

public class JugadorEspeculador extends Jugador
{
    private static int FactorEspeculador=2;
    private float fianza;

    JugadorEspeculador(Jugador jugador, int fianza)
    {
        super(jugador);
        this.fianza=fianza;
        this.especulador=true;

        for(int i=0; i<jugador.getPropiedades().size(); i++)
        {
            jugador.getPropiedades().get(i).actualizaPropietarioPorConversion(this);
        }
    }

    @Override
    protected int getCasasMax()
    {
        return super.getCasasMax()*JugadorEspeculador.FactorEspeculador;
    }

    @Override
    boolean pagaImpuesto(float cantidad)
    {
        return super.paga(cantidad/2);
    }

    @Override
    protected int getHotelesMax() 
    {
        return super.getHotelesMax()*JugadorEspeculador.FactorEspeculador;
    }

    @Override
    public String toString() {

        String salida;
        salida = "Nombre: " + this.getNombre() + "\n" + "Saldo: " + this.getSaldo() + "\n" + "Casilla actual: "
                + this.getNumCasillaActual() + "\n" + "Encarcelado: " + this.isEncarcelado() + "\n" + "Este jugador es un jugador especulador.";

        return salida;
    }
    
    @Override
    protected boolean debeSerEncarcelado() 
    {
        boolean carcel = false;

        if (!this.encarcelado) {
            if (!this.tieneSalvoConducto() && getSaldo() <= this.fianza)
                carcel = true;
            else if(this.tieneSalvoConducto())
                perderSalvoConducto();
            else
                paga(fianza);
        }

        return carcel;
    }
}